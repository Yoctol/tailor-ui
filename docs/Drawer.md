---
id: drawer
title: Drawer
---

Panel slides from screen edge.

## When To Use

A Drawer is a panel that is typically overlaid on top of a page and slides in from the side. It contains a set of information or actions. Since that user can interact with the Drawer without leaving the current page, tasks can be achieved more efficient within the same context.

- Use a Form to create or edit a set of information.
- Processing subtasks. When subtasks are too heavy for Popover and we still want to keep the subtasks in the context of the main task, Drawer - comes very handy.

When a same Form is needed in multiple places.

## Examples

```js
import { Drawer } from 'tailor-ui';
```

### Basic

```jsx live
() => {
  const [visible, setVisible] = useState(false);

  return (
    <>
      <Button onClick={() => setVisible(!visible)}>Open</Button>
      <Drawer title="Title" visible={visible} onClose={() => setVisible(false)}>
        {Array.from({ length: 100 }).map((_, index) => (
          <div key={index}>This is the content of Drawer</div>
        ))}
      </Drawer>
    </>
  );
}
```

### Custom Placement

```jsx live
() => {
  const [visible, setVisible] = useState(false);
  const [placement, setPlacement] = useState('left');

  return (
    <>
      <RadioGroup value={placement} onChange={setPlacement}>
        <Radio value="top">Top</Radio>
        <Radio value="right">Right (default)</Radio>
        <Radio value="bottom">Bottom</Radio>
        <Radio value="left">Left</Radio>
      </RadioGroup>

      <Button mt="3" onClick={() => setVisible(!visible)}>
        Open
      </Button>

      <Drawer
        title="Title"
        visible={visible}
        placement={placement}
        onClose={() => setVisible(false)}
      >
        {Array.from({ length: 100 }).map((_, index) => (
          <div key={index}>This is the content of Drawer</div>
        ))}
      </Drawer>
    </>
  );
}
```

### Add Custom Footer

```jsx live
() => {
  const [visible, setVisible] = useState(false);

  return (
    <>
      <Button onClick={() => setVisible(!visible)}>Open</Button>
      <Drawer
        title="Title"
        visible={visible}
        onClose={() => setVisible(false)}
        footer={
          <Button variant="primary" onClick={() => setVisible(false)}>
            Close Drawer
          </Button>
        }
      >
        {Array.from({ length: 100 }).map((_, index) => (
          <div key={index}>This is the content of Drawer</div>
        ))}
      </Drawer>
    </>
  );
}
```

### hidden close button and prevent mask clicked close

```jsx live
() => {
  const [visible, setVisible] = useState(false);

  return (
    <>
      <Button onClick={() => setVisible(!visible)}>Open</Button>
      <Drawer
        title="Title"
        visible={visible}
        onClose={() => setVisible(false)}
        closable={false}
        maskClosable={false}
        footer={
          <Button variant="primary" onClick={() => setVisible(false)}>
            Close Drawer
          </Button>
        }
      >
        {Array.from({ length: 100 }).map((_, index) => (
          <div key={index}>This is the content of Drawer</div>
        ))}
      </Drawer>
    </>
  );
}
```

## API

| Property     | Description | Type                                           | Default   |
| ------------ | ----------- | ---------------------------------------------- | --------- |
| visible      |             | `boolean`                                      |           |
| onClose      |             | `() => void`                                   |           |
| title        |             | `ReactNode`                                    |           |
| footer       |             | `ReactNode`                                    |           |
| placement    |             | `'top'` \| `'right'` \| `'bottom'` \| `'left'` | `'right'` |
| closable     |             | `boolean`                                      | `true`    |
| maskClosable |             | `boolean`                                      | `true`    |
| width        |             | `string` \| `number`                           | 400       |
| height       |             | `string` \| `number`                           | 300       |
