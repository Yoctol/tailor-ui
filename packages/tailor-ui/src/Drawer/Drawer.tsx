import React, {
  Fragment,
  FunctionComponent,
  ReactNode,
  useEffect,
} from 'react';
import styled, { css } from 'styled-components';
import { animated, config, useTransition } from 'react-spring';

import Backdrop from '../Backdrop';
import Portal from '../utils/Portal';
import tag from '../utils/CleanTag';
import useKeydown, { ESC_KEY_CODE } from '../utils/useKeydown';
import { FooterWrapper, ModalContent, ModalHeader } from '../Modal';

type Placement = 'top' | 'right' | 'bottom' | 'left';

interface DrawerWrapperProps {
  breadth: string;
  placement: Placement;
}

const DrawerWrapper = styled(tag.div)<DrawerWrapperProps>`
  display: flex;
  position: absolute;
  z-index: 10000;
  flex-direction: column;
  background-color: #fff;

  ${p =>
    ['top', 'bottom'].includes(p.placement)
      ? css`
          top: ${p.placement === 'top' ? 0 : `calc(100vh - ${p.breadth})`};
          height: ${p.breadth};
          width: 100vw;
          box-shadow: 0 ${p.placement === 'bottom' && '-'}2px 8px
            rgba(0, 0, 0, 0.15);
        `
      : css`
          left: ${p.placement === 'left' ? 0 : `calc(100vw - ${p.breadth})`};
          width: ${p.breadth};
          height: 100vh;
          box-shadow: ${p.placement === 'right' && '-'}2px 0 8px
            rgba(0, 0, 0, 0.15);
        `}
`;

const formatBreadth = (breadth: number | string) =>
  typeof breadth === 'number' ? `${breadth}px` : breadth;

const getTransformAxis = (placement: Placement) => {
  if (['top', 'bottom'].includes(placement)) {
    return 'translateY';
  }

  return 'translateX';
};

const getTransformBreadth = ({
  placement,
  breadth,
}: {
  placement: Placement;
  breadth: string;
}) => {
  const prefix = ['left', 'top'].includes(placement) ? '-' : '';
  return `${prefix}${breadth}`;
};

const getWrapperBreadth = ({
  placement,
  width,
  height,
}: {
  placement: Placement;
  width: number | string;
  height: number | string;
}) => {
  if (['top', 'bottom'].includes(placement)) {
    return formatBreadth(height);
  }

  return formatBreadth(width);
};

export interface DrawerProps {
  visible: boolean;
  onClose: () => void;
  title?: string;
  footer?: ReactNode;
  placement?: Placement;
  closable?: boolean;
  maskClosable?: boolean;
  width?: string | number;
  height?: string | number;
}

const Drawer: FunctionComponent<DrawerProps> = ({
  visible,
  title,
  footer,
  placement = 'right' as Placement,
  width = 400,
  height = 300,
  closable = true,
  maskClosable = true,
  onClose,
  children,
  ...otherProps
}) => {
  const breadth = getWrapperBreadth({ placement, width, height });
  const transformAxis = getTransformAxis(placement);
  const transformBreadth = getTransformBreadth({ placement, breadth });

  const transition = useTransition(visible, null, {
    from: {
      offset: transformBreadth,
    },
    enter: {
      offset: '0px',
    },
    leave: {
      offset: transformBreadth,
    },
    config: {
      ...config.default,
      precision: 0.1,
    },
  });

  useKeydown({
    listening: visible,
    keyCode: ESC_KEY_CODE,
    onKeydown: onClose,
  });

  useEffect(() => {
    document.body.style.overflow = visible ? 'hidden' : '';

    return () => {
      document.body.style.overflow = '';
    };
  }, [visible]);

  return (
    <Portal appendFor="drawer">
      {transition.map(
        ({ item, key, props }) =>
          item && (
            <Fragment key={key}>
              <Backdrop
                visible={visible}
                onClick={() => {
                  if (maskClosable) {
                    onClose();
                  }
                }}
              />
              <animated.div
                style={{
                  transform:
                    props.offset &&
                    props.offset.interpolate(
                      offset => `${transformAxis}(${offset})`
                    ),
                }}
              >
                <DrawerWrapper placement={placement} breadth={breadth}>
                  {title && (
                    <ModalHeader
                      title={title}
                      onCancel={onClose}
                      closable={closable}
                    />
                  )}
                  <ModalContent {...otherProps}>{children}</ModalContent>
                  {footer && <FooterWrapper>{footer}</FooterWrapper>}
                </DrawerWrapper>
              </animated.div>
            </Fragment>
          )
      )}
    </Portal>
  );
};

export default Drawer;
