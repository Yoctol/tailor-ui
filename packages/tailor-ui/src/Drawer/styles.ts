import styled, { css } from 'styled-components';

import { Placement } from './types';

interface DrawerWrapperProps {
  breadth: string;
  placement: Placement;
}

export const DrawerWrapper = styled.div<DrawerWrapperProps>`
  display: flex;
  position: absolute;
  z-index: 10000;
  flex-direction: column;
  padding: ${p => p.theme.space[3]};
  background-color: #fff;

  ${p =>
    ['top', 'bottom'].includes(p.placement)
      ? css`
          top: ${p.placement === 'top' ? 0 : `calc(100vh - ${p.breadth})`};
          height: ${p.breadth};
          width: 100vw;
          box-shadow: 0 ${p.placement === 'bottom' && '-'}2px 8px
            rgba(0, 0, 0, 0.15);
        `
      : css`
          left: ${p.placement === 'left' ? 0 : `calc(100vw - ${p.breadth})`};
          width: ${p.breadth};
          height: 100vh;
          box-shadow: ${p.placement === 'right' && '-'}2px 0 8px
            rgba(0, 0, 0, 0.15);
        `}
`;
