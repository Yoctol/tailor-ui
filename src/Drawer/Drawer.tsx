import React, { FunctionComponent, ReactNode, useEffect } from 'react';
import { Transition, animated, config } from 'react-spring';

import Portal from 'utils/Portal';
import styled, { css } from 'utils/styled-components';
import tag from 'utils/CleanTag';
import useKeydown, { ESC_KEY_CODE } from 'utils/useKeydown';

import Backdrop from '../Backdrop';
import { FooterWrapper, ModalContent, ModalHeader } from '../Modal';

type Placement = 'top' | 'right' | 'bottom' | 'left';

interface IDrawerWrapperProps {
  breadth: string;
  placement: Placement;
}

const DrawerWrapper = styled<IDrawerWrapperProps, 'div'>(tag.div)`
  display: flex;
  position: absolute;
  z-index: 10000;
  flex-direction: column;
  background-color: #fff;

  ${p =>
    ['top', 'bottom'].includes(p.placement)
      ? css`
          top: ${p.placement === 'top' ? 0 : `calc(100vh - ${p.breadth})`};
          height: ${p.breadth};
          width: 100vw;
          box-shadow: 0 ${p.placement === 'bottom' && '-'}2px 8px
            rgba(0, 0, 0, 0.15);
        `
      : css`
          left: ${p.placement === 'left' ? 0 : `calc(100vw - ${p.breadth})`};
          width: ${p.breadth};
          height: 100vh;
          box-shadow: ${p.placement === 'right' && '-'}2px 0 8px
            rgba(0, 0, 0, 0.15);
        `}
`;

const formatBreadth = (breadth: number | string) =>
  typeof breadth === 'number' ? `${breadth}px` : breadth;

const getTransformAxis = (placement: Placement) => {
  if (['top', 'bottom'].includes(placement)) {
    return 'translateY';
  }

  return 'translateX';
};

const getTransformBreadth = ({
  placement,
  breadth,
}: {
  placement: Placement;
  breadth: string;
}) => {
  const prefix = ['left', 'top'].includes(placement) ? '-' : '';
  return `${prefix}${breadth}`;
};

const getWrapperBreadth = ({
  placement,
  width,
  height,
}: {
  placement: Placement;
  width: number | string;
  height: number | string;
}) => {
  if (['top', 'bottom'].includes(placement)) {
    return formatBreadth(height);
  }

  return formatBreadth(width);
};

export interface IDrawerProps {
  visible: boolean;
  onClose: () => void;
  title?: string;
  footer?: ReactNode;
  placement?: Placement;
  closable?: boolean;
  maskClosable?: boolean;
  width?: string | number;
  height?: string | number;
}

const Drawer: FunctionComponent<IDrawerProps> = ({
  visible,
  title,
  footer,
  placement = 'right',
  width = 400,
  height = 300,
  closable = true,
  maskClosable = true,
  onClose,
  children,
  ...otherProps
}) => {
  const breadth = getWrapperBreadth({ placement, width, height });
  const transformAxis = getTransformAxis(placement);
  const transformBreadth = getTransformBreadth({ placement, breadth });

  useKeydown({
    listening: visible,
    keyCode: ESC_KEY_CODE,
    onKeydown: onClose,
  });

  useEffect(
    () => {
      document.body.style.overflow = visible ? 'hidden' : '';

      return () => {
        document.body.style.overflow = '';
      };
    },
    [visible]
  );

  return (
    <Portal appendFor="drawer">
      <Transition
        native
        items={visible}
        from={{
          offset: transformBreadth,
        }}
        enter={{
          offset: '0px',
        }}
        leave={{
          offset: transformBreadth,
        }}
        config={{
          ...config.default,
          precision: 0.1,
        }}
      >
        {_visible =>
          _visible &&
          ((props: { offset: any }) => (
            <>
              <Backdrop
                visible={visible}
                onClick={() => {
                  if (maskClosable) {
                    onClose();
                  }
                }}
              />
              <animated.div
                style={{
                  transform: props.offset.interpolate(
                    (offset: any) => `${transformAxis}(${offset})`
                  ),
                }}
              >
                <DrawerWrapper placement={placement} breadth={breadth}>
                  {title && (
                    <ModalHeader
                      title={title}
                      onCancel={onClose}
                      closable={closable}
                    />
                  )}
                  <ModalContent {...otherProps}>{children}</ModalContent>
                  {footer && <FooterWrapper>{footer}</FooterWrapper>}
                </DrawerWrapper>
              </animated.div>
            </>
          ))
        }
      </Transition>
    </Portal>
  );
};

export default Drawer;
