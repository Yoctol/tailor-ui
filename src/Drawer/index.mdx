---
name: Drawer
menu: Components
route: /components/drawer
---

import { PropsTable, Playground } from 'docz';

import Button from '../Button';
import Radio from '../Radio';

import Drawer from './';

# Drawer

Panel slides from screen edge.

## When To Use

A Drawer is a panel that is typically overlaid on top of a page and slides in from the side. It contains a set of information or actions. Since that user can interact with the Drawer without leaving the current page, tasks can be achieved more efficient within the same context.

- Use a Form to create or edit a set of information.
- Processing subtasks. When subtasks are too heavy for Popover and we still want to keep the subtasks in the context of the main task, Drawer - comes very handy.

When a same Form is needed in multiple places.

## Examples

```js
import { Drawer } from 'tailor-ui';
```

### Basic

<Playground>
  {() => {
    const [visible, setVisible] = React.useState(false);

    return (
      <>
        <Button onClick={() => setVisible(!visible)}>Open</Button>
        <Drawer
          title="Title"
          visible={visible}
          onClose={() => setVisible(false)}
        >
          Content...
        </Drawer>
      </>
    )

}}

</Playground>

### Custom Placement

<Playground>
  {() => {
    const [visible, setVisible] = React.useState(false);
    const [placement, setPlacement] = React.useState('left');

    return (
      <>
        <Radio.Group value={placement} onChange={setPlacement}>
          <Radio value="top">Top</Radio>
          <Radio value="right">Right (default)</Radio>
          <Radio value="bottom">Bottom</Radio>
          <Radio value="left">Left</Radio>
        </Radio.Group>

        <Button mt="2" onClick={() => setVisible(!visible)}>Open</Button>

        <Drawer
          title="Title"
          visible={visible}
          placement={placement}
          onClose={() => setVisible(false)}
        >
          Content...
        </Drawer>
      </>
    )

}}

</Playground>

### Add Custom Footer

<Playground>
  {() => {
    const [visible, setVisible] = React.useState(false);

    return (
      <>
        <Button onClick={() => setVisible(!visible)}>Open</Button>
        <Drawer
          title="Title"
          visible={visible}
          onClose={() => setVisible(false)}
          footer={
            <Button type="primary" onClick={() => setVisible(false)}>Close Drawer</Button>
          }
        >
          Content...
        </Drawer>
      </>
    )

}}

</Playground>

### hidden close button and prevent mask clicked close

<Playground>
  {() => {
    const [visible, setVisible] = React.useState(false);

    return (
      <>
        <Button onClick={() => setVisible(!visible)}>Open</Button>
        <Drawer
          title="Title"
          visible={visible}
          onClose={() => setVisible(false)}
          closable={false}
          maskClosable={false}
          footer={
            <Button type="primary" onClick={() => setVisible(false)}>Close Drawer</Button>
          }
        >
          Content...
        </Drawer>
      </>
    )

}}

</Playground>

## API

<PropsTable of={Drawer} />
