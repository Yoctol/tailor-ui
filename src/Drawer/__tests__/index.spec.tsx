import React from 'react';

import { render } from 'test/test-utils';

import Drawer from '..';

describe('Modal', () => {
  it('should render correctly', () => {
    const { container } = render(<Drawer visible onClose={() => {}} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with props closable', () => {
    const { container } = render(
      <Drawer visible closable={false} onClose={() => {}} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
