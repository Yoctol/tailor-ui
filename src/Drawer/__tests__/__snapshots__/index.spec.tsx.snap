// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render correctly 1`] = `null`;

exports[`Modal should render with props closable 1`] = `null`;
